core/types/typePrinter.ml

```diff
@@ -0,0 +1,928 @@
+    presence empty_context (policy (), tyvar_name_map) f
+
+  let pol_string_of_type_arg ?(policy = default_policy)
+      ?(refresh_tyvar_names = true) (arg : type_arg) =
+    let policy = policy () in
+    build_tyvar_names ~refresh_tyvar_names
+      FreeTypeVars.free_bound_type_arg_type_vars [arg] ;
+    let context =
+      context_with_shared_effect policy (fun o -> o#type_arg arg)
+    in
+    type_arg context (policy, tyvar_name_map) arg
+
+  let pol_string_of_row_var ?(policy = default_policy)
+      ?(refresh_tyvar_names = true) the_row_var =
+    build_tyvar_names ~refresh_tyvar_names FreeTypeVars.free_bound_row_var_vars
+      [the_row_var] ;
+    match
+      row_var name_of_type "," empty_context
+        (policy (), tyvar_name_map)
+        the_row_var
+    with
+    | None -> ""
+    | Some s -> s
+
+  let pol_string_of_tycon_spec ?(policy = default_policy)
+      ?(refresh_tyvar_names = true) (tycon : tycon_spec) =
+    build_tyvar_names ~refresh_tyvar_names
+      FreeTypeVars.free_bound_tycon_type_vars [tycon] ;
+    tycon_spec empty_context (policy (), tyvar_name_map) tycon
+
+  let pol_string_of_quantifier ?(policy = default_policy)
+      ?(refresh_tyvar_names = true) (quant : Quantifier.t) =
+    build_tyvar_names ~refresh_tyvar_names
+      FreeTypeVars.free_bound_quantifier_vars [quant] ;
+    quantifier (policy (), tyvar_name_map) quant
+
+  let string_of_datatype t = pol_string_of_datatype t
+  let string_of_row r = pol_string_of_row r
+  let string_of_field_spec p = pol_string_of_presence p
+  let string_of_type_arg ta = pol_string_of_type_arg ta
+  let string_of_quantifier q = pol_string_of_quantifier q
+  let string_of_tycon_spec ts = pol_string_of_tycon_spec ts
+
+  let mk_pp_fn fmt string_of_fn t = Format.pp_print_string fmt (string_of_fn t)
+
+  let pp_datatype : Format.formatter -> datatype -> unit =
+   fun fmt -> mk_pp_fn fmt string_of_datatype
+
+  let pp_row : Format.formatter -> row -> unit =
+   fun fmt -> mk_pp_fn fmt string_of_row
+
+  let pp_quantifier : Format.formatter -> Quantifier.t -> unit =
+   fun fmt -> mk_pp_fn fmt string_of_quantifier
+
+  let pp_type_arg : Format.formatter -> type_arg -> unit =
+   fun fmt -> mk_pp_fn fmt string_of_type_arg
+end
+
+module PrettyWithPolicy = struct
+  type policy = Pretty.policy =
+    { quantifiers: bool
+    ; flavours: bool
+    ; hide_fresh: bool
+    ; kinds: string
+    ; effect_sugar: bool }
+
+  let default_policy = Pretty.default_policy
+
+  let build_tyvar_names = Pretty.build_tyvar_names
+  let add_tyvar_names = Pretty.add_tyvar_names
+
+  let string_of_datatype = Pretty.pol_string_of_datatype
+  let string_of_row = Pretty.pol_string_of_row
+  let string_of_field_spec = Pretty.pol_string_of_presence
+  let string_of_type_arg = Pretty.pol_string_of_type_arg
+  let string_of_tycon_spec = Pretty.pol_string_of_tycon_spec
+  let string_of_row_var = Pretty.pol_string_of_row_var
+end
+
+module BySetting =
```

Comment:
I just realized that the previous definition (using first-class modules) for defining `BySetting` was too eager: The old version would not have updated the behavior if the user changes `print_types_pretty` via the CLI.

However, the new version of `BySetting` is very repetitive. Any ideas for a nicer solution?